### PR TITLE
Support dynamic OCP default version

### PR DESCRIPTION
**Pull request by YuviGold**
Return the openshift version that needs to be handled
according to the following process:

1. In case env var `OPENSHIFT_INSTALL_RELEASE_IMAGE` is defined to override the release image -
extract its OCP version.
2. In case env var `OPENSHIFT_VERSION` is defined - return it.
3. In case of `allow_default` is enabled, return the default supported version by assisted-service.
3.1 If a client is provided, request the versions from the service (supports remote service).
3.2 Otherwise, Get from the JSON file in assisted-service repository.

/cc @eliorerz

### CODE REVIEW

File: discovery-infra/test_infra/utils/utils.py [hunk 1]
```diff
@@ -685,7 +685,14 @@ def get_kubeconfig_path(cluster_name: str) -> str:
     return kubeconfig_path
 
 
-def get_openshift_version(default=consts.DEFAULT_OPENSHIFT_VERSION):
+def get_default_openshift_version() -> str:
+    stdout, _, _ = run_command(
+        f"jq -r 'to_entries | .[] | select(.value.default) | .key' assisted-service/data/default_ocp_versions.json", shell=True
```
Comment (tsorya) [inline discovery-infra/test_infra/utils/utils.py:690]: if you run test-infra on remote-url, without spawning minikube and deploying service, you will not get this fille.
I think you should validate it before.
Why to use jq? You are using python , just json.load it

Comment (osherdp) [inline discovery-infra/test_infra/utils/utils.py:690]: just use python and ``json.load``

File: discovery-infra/test_infra/utils/utils.py [hunk 2]
```diff
@@ -697,13 +704,20 @@ def get_openshift_version(default=consts.DEFAULT_OPENSHIFT_VERSION):
             )
         return stdout
 
-    return get_env("OPENSHIFT_VERSION", default)
+    version = get_env("OPENSHIFT_VERSION")
+    if version:
+        return version
 
+    if allow_default:
+        return get_default_openshift_version()
```
Comment (osherdp) [inline discovery-infra/test_infra/utils/utils.py:713]: at the end of the function, return ``None`` explicitly

**Review by osherdp — CHANGES_REQUESTED** [review]

File: discovery-infra/test_infra/utils/utils.py (hunk 2 above)
Comment (YuviGold) [reply discovery-infra/test_infra/utils/utils.py:713]: why? In case no return statement, `None` is returned

Comment (osherdp) [reply discovery-infra/test_infra/utils/utils.py:713]: I know, that's why I said explicit
That's the styling choice that I prefer

File: discovery-infra/test_infra/utils/utils.py (hunk 1 above)
Comment (YuviGold) [reply discovery-infra/test_infra/utils/utils.py:690]: That's true. So maybe we should use the client versions API instead.

**Comment by YuviGold** [comment]
/hold
In order to support remote clusters, it would be better to fetch the versions from the client.

File: discovery-infra/tests/conftest.py [hunk 3]
```diff
@@ -13,25 +12,14 @@ def api_client():
     yield get_api_client()
 
 
-def get_api_client(offline_token=None, **kwargs) -> InventoryClient:
```
[the remark this answers is not in the document]
Comment (YuviGold) [reply discovery-infra/tests/conftest.py:16]: https://gitlab.cee.redhat.com/ocp-edge-qe/kni-assisted-installer-auto/-/merge_requests/476

**Comment by YuviGold** [comment]
Everything passed :tada: 
/unhold

File: discovery-infra/test_infra/utils/global_variables/env_variables_utils.py [hunk 4]
```diff
@@ -16,7 +15,7 @@ class _EnvVariablesUtils(ABC):
     remote_service_url: str = get_env("REMOTE_SERVICE_URL")
     pull_secret: str = get_env("PULL_SECRET")
     offline_token: str = get_env("OFFLINE_TOKEN")
-    openshift_version: str = get_openshift_version()
+    openshift_version: str = ''
```
Comment (tsorya) [inline discovery-infra/test_infra/utils/global_variables/env_variables_utils.py:18]: maybe None?

File: discovery-infra/test_infra/utils/global_variables/global_variables.py [hunk 5]
```diff
@@ -44,6 +47,11 @@ class GlobalVariables(_EnvVariablesUtils):
 
     def __post_init__(self):
         super().__post_init__()
+        client=None
+        if not self.is_kube_api:
+            with suppress(RuntimeError, TimeoutError):
```
Comment (tsorya) [inline discovery-infra/test_infra/utils/global_variables/global_variables.py:52]: why it is ok to fail ? please comment

File: discovery-infra/test_infra/utils/utils.py [hunk 6]
```diff
@@ -740,7 +735,31 @@ def get_kubeconfig_path(cluster_name: str) -> str:
     return kubeconfig_path
 
 
-def get_openshift_version(default=consts.DEFAULT_OPENSHIFT_VERSION):
+def get_default_openshift_version(client=None) -> str:
```
Comment (tsorya) [inline discovery-infra/test_infra/utils/utils.py:738]: not sure regarding utils. Why it is not part of cluster configuration classes?

**Pull request merged**